runtime/bus.go

```diff
@@ -45,7 +45,7 @@ func (m *bus) GetModulesRegistry() modules.ModulesRegistry {
 	return m.modulesRegistry
 }
 
-func (m *bus) RegisterModule(module modules.Module) {
+func (m *bus) RegisterModule(module modules.Submodule) {
```

Comment:
This is actually a bit confusing to me.

A `submodule` is a strict subset of a submodule, so do we not want this to be a `Module`?

Reply:
I believe this needs to be a `Submodule` for submodules to call `bus.RegisterModule()` and be able to be retrived as `bus.GetXYZSubmodule()` the `Start()` and `Stop()` methods etc in the `Module` interface that are not in the `Submodule` interface constrain their ability to be added to the bus. But the reverse is still functional as the `Module` contains the `Submodule` interface functions `GetModuleName()` and `GetBus()` `SetBus()` we can pass modules into this without any typecasting etc for the same current functionality.

Reply:
@h5law's correct, `Module` is overly constraining for the modules registry. While we have the `Integrable` base module to work around the  `Start()` and `Stop()` implementation mentioned, the main over-constraint is on `Module#Create()`.

The modules registry is only concerned with the smallest interface, `InjectableModule`. The bus methods which interact with the modules registry imply that the registered (sub)modules are also `IntegrableModule`s (i.e. hold a reference to the bus). The union of these two interfaces is the definition of `Submodule`.

This is the most significant distinction between modules and submodules in my view. The consistency and simplicity afforded by the `Module` interface is perfect for that level of abstraction but it's become clear that modules have dependencies which need to be dependency-injectable but would require unnecessary additional complexity to be interoperable with the modules registry.